### PR TITLE
Testing: Add an e2e test to check the interactions in write mode

### PR DESCRIPTION
**Pull request by youknowriad**
Related #60021 

## What?

This PR adds a first e2e test to check the different interactions in the "write" mode. 
Starting with inspector and mouse selection.

### CODE REVIEW

File: packages/e2e-test-utils-playwright/src/editor/switch-editor-tool.ts
Comment (kevin940726) [inline packages/e2e-test-utils-playwright/src/editor/switch-editor-tool.ts]: I believe this was not updated?

File: packages/e2e-test-utils-playwright/src/editor/switch-editor-tool.ts
Comment (kevin940726) [inline packages/e2e-test-utils-playwright/src/editor/switch-editor-tool.ts:12]: Nit: Prefer using unions of string literals in types since we only have two options for now. Something like `'Design' | 'Write'`.

File: test/e2e/specs/editor/various/write-design-mode.spec.js
Comment (kevin940726) [inline test/e2e/specs/editor/various/write-design-mode.spec.js]: Could we use `toHaveText` or `toContainText` here?

File: test/e2e/specs/editor/various/write-design-mode.spec.js
Comment (kevin940726) [inline test/e2e/specs/editor/various/write-design-mode.spec.js:110]: Using class name selectors is not recommended. Could we instead select it by `getByRole` and optionally with a `level` field here?

Comment (youknowriad) [reply test/e2e/specs/editor/various/write-design-mode.spec.js:110]: I know but for me, this is not easy to pinpoint and avoid random failures when we add headings... I prefer the selector here.

File: packages/e2e-test-utils-playwright/src/editor/switch-editor-tool.ts
Comment (youknowriad) [reply packages/e2e-test-utils-playwright/src/editor/switch-editor-tool.ts:12]: I'm hesitant, someone could decide to test another language.

File: test/e2e/specs/editor/various/write-design-mode.spec.js
Comment (kevin940726) [reply test/e2e/specs/editor/various/write-design-mode.spec.js:110]: I think there's already a heading in place? What do you mean by pinpointing random failures? I'd appreciate if we could follow the established [best practices](https://github.com/WordPress/gutenberg/tree/trunk/docs/contributors/code/e2e#use-accessible-selectors) whenever possible and add comments if we could not. Thanks! 🙏

Comment (youknowriad) [reply test/e2e/specs/editor/various/write-design-mode.spec.js:110]: The `h2` is not enough to say that I'm targeting the "title" of the inspector panel. There could be more h2 in the inspector panel, in fact there are already. (content).

We could theoretically say `h2 > eq( 0 )` but that's not very stable, any random design change can break that. (moving titles around).

So for me, in this case, I prefer a non ambiguous assertion rather than a generic one. 

I agree on following best practices, but we need to understand what's important and why they exist. 

I'm happy to update If there's a less ambiguous approach using that doesn't use selectors.

Comment (kevin940726) [reply test/e2e/specs/editor/various/write-design-mode.spec.js:110]: A common workaround practice throughout the codebase is using `expect( getByRole('heading', {name: 'Group', level: 2}) ).toBeVisible()`.

Comment (youknowriad) [reply test/e2e/specs/editor/various/write-design-mode.spec.js:110]: That works, but I'm not sure it's good.

 - I want to test that the block card title's is "group"
 - I don't want to test that there's an h2 who's content is "group".

Comment (kevin940726) [reply test/e2e/specs/editor/various/write-design-mode.spec.js:110]: The problem is that for end users there's no way to connect `.block-editor-block-card__title` to the block card title, they are not even aware of the class name. A simple way to think of this is from the perspective of screen readers. What screen readers read is what we test here. If we want to bring more context to the headings we should make that information accessible to them too.

File: packages/e2e-test-utils-playwright/src/editor/switch-editor-tool.ts
Comment (kevin940726) [reply packages/e2e-test-utils-playwright/src/editor/switch-editor-tool.ts:12]: They're only types so I guess they can always bypass that. We can also add `| string` to the end to allow arbitrary strings.

File: test/e2e/specs/editor/various/write-design-mode.spec.js
Comment (youknowriad) [reply test/e2e/specs/editor/various/write-design-mode.spec.js:110]: That makes a lot of sense I agree. How can we do that?

Comment (kevin940726) [reply test/e2e/specs/editor/various/write-design-mode.spec.js:110]: Let's stick with the current practice of `getByRole` and figure out the accessibility concern in a new issue?

Comment (getdave) [reply test/e2e/specs/editor/various/write-design-mode.spec.js:110]: For me, when this type of issue arrises it's an indicator that the a11y of the code in question could be improved.

Riad is correct the `getByRole` is (in this particular instance) ambiguous and could be fragile.

Kai is right that we should strive to interact with perceivable elements rather than rely on "developer APIs" such as CSS classnames. 

Can we improve the markup here to allow the selector to be less ambiguous? For me it feels like the name of the block is the title for the _entire_ sidebar so it should be a higher heading level than those used for the panels below. That would indicate a hierarchy, where the block controls themselves are "owned" by the "block".

We could also consider using [`<hgroup>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/hgroup)

> The <hgroup> element represents a heading and related content. 

Or even `<header>` if we introduced a sectioning level element.

Comment (youknowriad) [reply test/e2e/specs/editor/various/write-design-mode.spec.js:110]: I feel like until we fix a potential a11y thing here, we should stick with the class for the test as it's more solid.

Comment (youknowriad) [reply test/e2e/specs/editor/various/write-design-mode.spec.js:110]: On the a11y approach, I think it's less about heading levels, and more about having an explicit aria-label or something like `Selected block type is "something"` but I may be wrong there and welcome alternatives.

Comment (ntsekouras) [reply test/e2e/specs/editor/various/write-design-mode.spec.js:110]: I'd echo Riad about until fixing the a11y issue, we should stick with that classname for now and unblock this PR. Let's create an issue for this.

**Review by ntsekouras — APPROVED** [review]
LGTM, thanks!

We need these kind of tests sooner than later. Let's create an issue for [this](https://github.com/WordPress/gutenberg/pull/65819#discussion_r1787432531) and ship this.

File: test/e2e/specs/editor/various/write-design-mode.spec.js
Comment (getdave) [reply test/e2e/specs/editor/various/write-design-mode.spec.js:110]: No need to block this PR. But we should raise a followup Issue [as suggested](https://github.com/WordPress/gutenberg/pull/65819#discussion_r1787374700) with the appropriate labels to get Accessibility feedback.

Comment (kevin940726) [reply test/e2e/specs/editor/various/write-design-mode.spec.js:110]: IMO, `getByRole` is always preferred, in cases when it doesn't apply, we should still fallback to alternatives like `getByTestId` than class name selectors. I don't understand the argument of using class name selectors being more "solid". TBH, this has happened too many times that this is becoming frustrating to me. Could we at least add a comment above indicating that this is an edge case so that people won't accidentally follow the practice? I don't think it should block this PR, but I think this type of issue is still important, especially given that it just flagged a potential accessibility concern here by trying to avoid the class name selectors.

Comment (youknowriad) [reply test/e2e/specs/editor/various/write-design-mode.spec.js:110]: I'll add a comment. I'm not sure we used test ids before.

Comment (getdave) [reply test/e2e/specs/editor/various/write-design-mode.spec.js:110]: Thanks for the comment.

Nit: the wording isn't quite accurate and should read (emphasis added for clarity):

```
// Ideally we should **_NOT_** be using CSS selectors
```

File: test/e2e/specs/editor/various/write-design-mode.spec.js
Comment (kevin940726) [inline test/e2e/specs/editor/various/write-design-mode.spec.js:107]: "not" be using CSS selectors? 😅

Comment (youknowriad) [reply test/e2e/specs/editor/various/write-design-mode.spec.js:107]: haha :)

Comment (youknowriad) [reply test/e2e/specs/editor/various/write-design-mode.spec.js:107]: It's fixed in trunk.

**Pull request merged**
